src/util/gcs.clj

```diff
@@ -0,0 +1,81 @@
+(ns util.gcs
+  "Utility functions for Google Cloud Storage shared across this program."
+  (:require [clojure.pprint :refer [pprint]]
+            [clojure.data.json  :as json]
+            [clojure.string     :as str]
+            [clojure.java.io    :as io]
+            [clj-http.client    :as http]
+            [vault.client.http]
+            [ptc]
+            [util.once          :as once])
+  (:import  [org.apache.tika Tika]))
+
+(def api-url
+  "The Google Cloud API URL."
+  "https://www.googleapis.com/")
+
+(def storage-url
+  "The Google Cloud URL for storage operations."
+  (str api-url "storage/v1/"))
+
+(def bucket-url
+  "The Google Cloud Storage URL for bucket operations."
+  (str storage-url "b/"))
+
+(def upload-url
+  "The Google Cloud Storage URL for upload operations."
+  (str api-url "upload/storage/v1/b/"))
+
+(defn parse-gs-url
+  "Return BUCKET and OBJECT from a gs://bucket/object URL."
+  [url]
+  (let [[gs-colon nada bucket object] (str/split url #"/" 4)]
+    (when-not
+     (and (every? seq [gs-colon bucket])
+          (= "gs:" gs-colon)
+          (= "" nada))
+      (throw (IllegalArgumentException. (format "Bad GCS URL: '%s'" url))))
+    [bucket (or object "")]))
+
+(defn list-objects
+  "The objects in BUCKET with PREFIX in a lazy sequence."
+  ([bucket prefix]
+   (letfn [(each [pageToken]
+             (let [{:keys [items nextPageToken]}
+                   (-> {:method       :get   ;; :debug true :debug-body true
+                        :url          (str bucket-url bucket "/o")
+                        :content-type :application/json
+                        :headers      (once/get-auth-header!)
+                        :query-params {:prefix prefix
+                                       :maxResults 999
+                                       :pageToken pageToken}}
+                       http/request
+                       :body
+                       (json/read-str :key-fn keyword))]
+               (lazy-cat items (when nextPageToken (each nextPageToken)))))]
+     (each "")))
+  ([bucket]
+   (list-objects bucket "")))
+
+(defn upload-file
+  "Upload FILE to BUCKET with name OBJECT."
+  ([file bucket object headers]
+   (let [body (io/file file)]
+     (-> {:method       :post  ;; :debug true :debug-body true
+          :url          (str upload-url bucket "/o")
+          :query-params {:uploadType "media"
+                         :name       object}
+          :content-type (.detect (new Tika) body)
+          :headers      headers
+          :body         body}
+         http/request
+         :body
+         (json/read-str :key-fn keyword))))
+  ([file bucket object]
+   (upload-file file bucket object (once/get-auth-header!)))
+  ([file url]
+   (apply upload-file file (parse-gs-url url))))
+
+(comment
+  (list-objects "broad-gotc-dev-zero-test")
+  (upload-file "project.clj" "broad-gotc-dev-zero-test" "junk-test.clj"))
```

Comment:
I think these can be turned into unit tests!

Reply:
I haven't found a good way to mock the GCS without using https://github.com/googleapis/google-cloud-java/tree/master/google-cloud-clients/google-cloud-contrib/google-cloud-nio with GCloud Java Clients, so i only adopted some of the existing unit tests from the `WFL` repo that don't require mocking. I think for this specific one, an integration test that works with an actual bucket might be easier for us, WDYT?

Reply:
done!